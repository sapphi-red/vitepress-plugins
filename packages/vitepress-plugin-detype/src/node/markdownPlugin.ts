import type MarkdownIt from 'markdown-it'
import { transform, removeMagicComments } from 'detype'
import type { PrettierOptions } from 'detype'
import type { ContentMap } from './contentMap'

const tabsShareStateKey = '~detype'
const langs = ['ts', 'js'] as const
const detypeInfoRE = /^ts({[^}]*})?,=detype({[^}]*})?=$/

export const detypePlugin = (
  md: MarkdownIt,
  prettierOptions: PrettierOptions,
  contentMap: ContentMap
) => {
  const tabLabelsProp = `:tabLabels="${md.utils.escapeHtml(
    JSON.stringify(langs)
  )}"`
  const shareStateKeyProp = `sharedStateKey="${md.utils.escapeHtml(
    tabsShareStateKey
  )}"`

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const originalFence = md.renderer.rules.fence!
  md.renderer.rules.fence = (...args) => {
    const [tokens, idx, , env] = args
    const token = tokens[idx]

    const m = token.info.trim().match(detypeInfoRE)
    if (!m) {
      return originalFence(...args)
    }

    const [, tsAttrs = '', jsAttrs = ''] = m
    const slots = langs.map(lang => {
      const attrs = lang === 'ts' ? tsAttrs : jsAttrs
      const output = (async () => {
        try {
          const content =
            lang === 'ts'
              ? removeMagicComments(token.content, 'foo.ts', prettierOptions)
              : await transform(token.content, 'foo.ts', {
                  prettierOptions,
                  removeTsComments: true
                })
          const codeWithFence =
            `${token.markup}${lang}${attrs}\n` + content + token.markup
          return { result: md.render(codeWithFence, env) }
        } catch (e) {
          return { error: e }
        }
      })()
      const key = contentMap.add(
        `${token.markup}_${lang}_${attrs}`,
        token.content,
        output
      )
      return `<template #${lang}>${key}</template>`
    })

    return `<PluginTabs ${tabLabelsProp} ${shareStateKeyProp}>${slots.join(
      ''
    )}</PluginTabs>`
  }
}
