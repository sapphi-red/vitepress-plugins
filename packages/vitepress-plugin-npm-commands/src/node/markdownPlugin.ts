import type MarkdownIt from 'markdown-it'
import { autoConverter } from './autoConverter'
import type { PackageManager } from './packageManager'
import { packageManagers } from './packageManager'

const tabsShareStateKey = '~npm-commands'
const npmCommandsCommandRE = new RegExp(
  `(?://\\s*\\[!=npm\\s+(${packageManagers.join('|')}|auto)\\])+$`,
  'm'
)

export const npmCommandsPlugin = (md: MarkdownIt) => {
  const shareStateKeyProp = `sharedStateKey="${md.utils.escapeHtml(
    tabsShareStateKey
  )}"`

  const originalFence = md.renderer.rules.fence!
  md.renderer.rules.fence = (...args) => {
    const [tokens, idx, , env] = args
    const token = tokens[idx]
    if (token.attrGet('----npm-disable') !== null) {
      // escape hatch to disable this feature
      const pos = token.attrIndex('----npm-disable')
      token.attrs?.splice(pos, 1)
      return originalFence(...args)
    }
    if (!npmCommandsCommandRE.test(token.content)) {
      return originalFence(...args)
    }

    const codes = generateEachPackageManagerCode(token.content)

    const slots = codes.map(([pkgManger, code]) => {
      const attrStr = token.attrs
        ? `{${token.attrs
            .map(([key, val]) => (val ? `${key}=${val}` : key))
            .join(' ')}}`
        : ''
      const codeWithFence =
        `${token.markup}${token.info}${attrStr}\n` + code + token.markup
      return `<PluginTabsTab label="${pkgManger}">${md.render(
        codeWithFence,
        env
      )}</PluginTabsTab>`
    })

    return `<PluginTabs ${shareStateKeyProp}>${slots.join('')}</PluginTabs>`
  }
}

const parseNpmCommandsCommandFromLine = (line: string) => {
  const m = line.match(npmCommandsCommandRE)
  if (!m) return null
  return m[1] as PackageManager | 'auto'
}

const generateEachPackageManagerCode = (input: string) => {
  const codes = Object.fromEntries(
    packageManagers.map(m => [m, [] as string[]])
  ) as Record<PackageManager, string[]>
  for (const line of input.split('\n')) {
    const pkgManager = parseNpmCommandsCommandFromLine(line)
    if (!pkgManager) {
      for (const key of packageManagers) {
        codes[key].push(line)
      }
    } else if (pkgManager === 'auto') {
      for (const key of packageManagers) {
        const convertedCommand = autoConverter(
          line.replace(npmCommandsCommandRE, '').trimEnd(),
          key
        )
        codes[key].push(convertedCommand)
      }
    } else {
      codes[pkgManager].push(line.replace(npmCommandsCommandRE, ''))
    }
  }
  return Object.entries(codes).map(
    ([key, val]) => [key as PackageManager, val.join('\n')] as const
  )
}
