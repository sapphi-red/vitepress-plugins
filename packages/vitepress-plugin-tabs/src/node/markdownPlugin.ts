import type MarkdownIt from 'markdown-it'
import container from 'markdown-it-container'
import type Token from 'markdown-it/lib/token'

type Params = {
  shareStateKey: string | undefined
}

const parseTabsParams = (input: string): Params => {
  const match = input.match(/key:(\S+)/)
  return {
    shareStateKey: match?.[1]
  }
}

export const tabsPlugin = (md: MarkdownIt) => {
  md.use(container, 'tabs', {
    render(tokens: Token[], index: number) {
      const token = tokens[index]
      if (token.nesting === 1) {
        const params = parseTabsParams(token.info)
        const shareStateKeyProp = params.shareStateKey
          ? `sharedStateKey="${md.utils.escapeHtml(params.shareStateKey)}"`
          : ''
        return `<PluginTabs ${shareStateKeyProp}>\n`
      } else {
        return `</PluginTabs>\n`
      }
    }
  })

  md.use(container, 'tab', {
    render(tokens: Token[], index: number) {
      const token = tokens[index]
      if (token.nesting === 1) {
        const label = token.info.replace(/^\s*tab\s*/, '')
        const labelProp = `label="${md.utils.escapeHtml(label)}"`
        return `<PluginTabsTab ${labelProp}>\n`
      } else {
        return `</PluginTabsTab>\n`
      }
    }
  })
}
