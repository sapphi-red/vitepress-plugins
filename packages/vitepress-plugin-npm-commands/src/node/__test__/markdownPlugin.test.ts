import MarkdownIt from 'markdown-it'
import { test, expect } from 'vitest'
import { npmCommandsPlugin } from '../markdownPlugin'
import dedent from 'ts-dedent'

const setup = () => {
  const md = MarkdownIt()
  md.use(npmCommandsPlugin)
  return md
}

const s = '```'

test('render manual correctly', () => {
  const md = setup()
  const result = md.render(dedent`
    ${s}
    npm i vite // [!=npm npm]
    yarn add vite // [!=npm yarn]
    pnpm add vite // [!=npm pnpm]
    bun add vite // [!=npm bun]
    ${s}
  `)
  expect(result).toMatchSnapshot()
})

test('render auto correctly', () => {
  const md = setup()
  const result = md.render(dedent`
    ${s}
    npm i // [!=npm auto]
    ${s}
  `)
  expect(result).toMatchSnapshot()
})

test('escape hatch', () => {
  const md = setup()
  const result = md.render(dedent`
    ${s}{----npm-disable}
    npm i // [!=npm auto]
    ${s}
  `)
  expect(result).toMatchSnapshot()
})
