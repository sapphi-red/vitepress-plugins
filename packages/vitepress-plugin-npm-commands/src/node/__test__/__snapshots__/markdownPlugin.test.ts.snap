// Vitest Snapshot v1

exports[`escape hatch 1`] = `
"<PluginTabs :tabLabels=\\"[&quot;npm&quot;,&quot;yarn&quot;,&quot;pnpm&quot;]\\" sharedStateKey=\\"~npm-commands\\"><template #npm><pre><code class=\\"language-{----npm-disable}\\">npm i
</code></pre>
</template><template #yarn><pre><code class=\\"language-{----npm-disable}\\">yarn install
</code></pre>
</template><template #pnpm><pre><code class=\\"language-{----npm-disable}\\">pnpm install
</code></pre>
</template></PluginTabs>"
`;

exports[`render auto correctly 1`] = `
"<PluginTabs :tabLabels=\\"[&quot;npm&quot;,&quot;yarn&quot;,&quot;pnpm&quot;]\\" sharedStateKey=\\"~npm-commands\\"><template #npm><pre><code>npm i
</code></pre>
</template><template #yarn><pre><code>yarn install
</code></pre>
</template><template #pnpm><pre><code>pnpm install
</code></pre>
</template></PluginTabs>"
`;

exports[`render manual correctly 1`] = `
"<PluginTabs :tabLabels=\\"[&quot;npm&quot;,&quot;yarn&quot;,&quot;pnpm&quot;]\\" sharedStateKey=\\"~npm-commands\\"><template #npm><pre><code>npm i vite 
</code></pre>
</template><template #yarn><pre><code>yarn add vite 
</code></pre>
</template><template #pnpm><pre><code>pnpm add vite 
</code></pre>
</template></PluginTabs>"
`;
